packages/actor-rdf-resolve-quad-pattern-tree/lib/ActorRdfResolveQuadPatternTree.ts

```diff
@@ -0,0 +1,147 @@
+
+    // Find a quad pattern or triple pattern search form
+    const searchForms: ISearchForms = metadata.searchForms;
+    if (!searchForms || !searchForms.values.length) {
+      throw new Error('No Hydra search forms were discovered in the metadata of ' + hypermedia
+        + '. You may be missing an actor that extracts this metadata');
+    }
+
+    // TODO: in the future, a query-based search form getter should be used.
+    for (const searchForm of searchForms.values) {
+      if (this.graphUri
+        && this.subjectUri in searchForm.mappings
+        && this.predicateUri in searchForm.mappings
+        && this.objectUri in searchForm.mappings
+        && this.graphUri in searchForm.mappings
+        && Object.keys(searchForm.mappings).length === 4) {
+        return searchForm;
+      }
+      if (this.subjectUri in searchForm.mappings
+        && this.predicateUri in searchForm.mappings
+        && this.objectUri in searchForm.mappings
+        && Object.keys(searchForm.mappings).length === 3) {
+        return searchForm;
+      }
+    }
+
+    throw new Error('No valid Hydra search form was found for quad pattern or triple pattern queries.');
+  }
+
+  protected async createSource(context: ActionContext): Promise<ILazyQuadSource> {
+    // Determine form lazily when a URL is constructed.
+    let chosenForm: Promise<ISearchForm> = null;
+
+    // Create a quad pattern to URL converter
+    const uriConstructor = async (subject?: RDF.Term, predicate?: RDF.Term, object?: RDF.Term, graph?: RDF.Term) => {
+      if (!chosenForm) {
+        // Collect metadata of the hypermedia
+        const hypermedia: string = this.getContextSource(context).value;
+
+        // Save the form, so it is determined only once per source.
+        chosenForm = this.chooseForm(hypermedia, context);
+      }
+
+      const entries: {[id: string]: string} = {};
+      const input = [
+        { uri: this.subjectUri, term: subject },
+        { uri: this.predicateUri, term: predicate },
+        { uri: this.objectUri, term: object },
+        { uri: this.graphUri, term: graph },
+      ];
+      for (const entry of input) {
+        if (entry.uri && entry.term) {
+          entries[entry.uri] = termToString(entry.term);
+        }
+      }
+
+      return (await chosenForm).getUri(entries);
+    };
+
+    return new MediatedQuadSource(this.mediatorRdfDereferencePaged, uriConstructor, context);
+  }
+
+  protected async getSource(context: ActionContext): Promise<ILazyQuadSource> {
+    // Cache the source object for each hypermedia entrypoint
+    const hypermedia: string = this.getContextSource(context).value;
+    if (this.sources[hypermedia]) {
+      return this.sources[hypermedia];
+    }
+
+    // Cache and return
+    return await (this.sources[hypermedia] = this.createSource(context));
+  }
+
+  protected async getOutput(source: RDF.Source, pattern: RDF.Quad, context: ActionContext)
+  : Promise<IActorRdfResolveQuadPatternOutput> {
+    debugger;
```

Comment:
debug code

Reply:
Oops, I'll pay attention that this doesn't get in here anymore.